### PR TITLE
VIRTS-1080: Gameboard Blue Visualization

### PR DESCRIPTION
**Pull request by mchan143**
- Links aligned by PID
- Scoreboard cleaned up, operation status moved next to operation select
- Red links are hidden for Access.BLUE, blue links are hidden for Access.RED
- Modal cleaned up

### CODE REVIEW

File: static/js/gameboard.js [hunk 1]
```diff
@@ -0,0 +1,237 @@
+$(document).ready(function () {
+    let redOps = $('#red-operations').children('option').length-1;
+    let blueOps = $('#blue-operations').children('option').length-1;
+    if(redOps===0 && blueOps===0) {
+        stream('GameBoard is only useful if a red or blue operation has been started.')
+    } else {
+        stream('Select red/blue operations to see what the defense detected and responded to.');
+    }
+});
+//let refresher = setInterval(refresh, 10000);
```
Comment (wbooth) [inline static/js/gameboard.js]: drop dead code

File: static/js/gameboard.js [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            document.getElementById('piece-modal').style.display='block';
+            let link = data.exchanges[id[1]][id[2]][id[3]];
+            $('#piece-cmd').html(atob(link['command']));
+            let factList = $('#piece-fact-list');
+            link['facts'].forEach(function(fact) {
+                let pieceFact = $('#piece-fact').clone();
+                pieceFact.html(fact.trait + ": " + fact.value);
+                pieceFact.show();
+                factList.append(pieceFact);
+//                getSuggestedQueries(fact);  // get suggested queries based on the passed in fact
```
Comment (wbooth) [inline static/js/gameboard.js]: drop dead code

File: static/js/gameboard.js [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+        $('#exchanges').prepend(piece);
+        piece.show();
+    }
+
+    $('#gb-blue-points').text(bluePoints);
+    $('#gb-red-points').text(redPoints);
+}
+
+function getSuggestedQueries(fact) {
```
Comment (wbooth) [inline static/js/gameboard.js]: drop for now, put in followup

**Comment by ghost** [comment]
@wbooth how close are we to getting some functionality out there for game board?

**Review by wbooth — APPROVED** [review]

File: app/gameboard_api.py [hunk 4]
```diff
@@ -14,17 +14,54 @@ def __init__(self, services):
...
+                        blue_op=blue_op[0].display if blue_op else None,
+                        exchanges=self._get_exchanges(red_op, blue_op))
         return web.json_response(response)
+
+    def _get_exchanges(self, red_ops, blue_ops):
+        exchanges = dict()
+        red_links = sorted([link for op in red_ops for link in op.chain if link.finish and link.cleanup == 0], key=lambda i: i.finish)
+        blue_links = sorted([link for op in blue_ops for link in op.chain if link.finish and link.cleanup == 0], key=lambda i: i.finish)
+        for link in red_links:
+            if link.pid in exchanges.keys():
```
Comment (ghost) [inline app/gameboard_api.py]: lot of duplicate code in this block. can we condense?

File: app/gameboard_api.py [hunk 5]
```diff
@@ -14,17 +14,58 @@ def __init__(self, services):
...
-        red_ops = await self.data_svc.locate('operations', dict(id=data.get('red')))
-        blue_ops = await self.data_svc.locate('operations', dict(id=data.get('blue')))
-        pieces = [l for op in red_ops + blue_ops for l in op.chain if l.finish]
-        response = dict(blue_state=blue_ops[0].state if blue_ops else 'n/a',
-                        red_state=red_ops[0].state if red_ops else 'n/a',
-                        links=[p.display for p in sorted(pieces, key=lambda i: i.finish)])
+        red_op = await self.data_svc.locate('operations', dict(id=data.get('red')))
+        blue_op = await self.data_svc.locate('operations', dict(id=data.get('blue')))
+        access = await self.auth_svc.get_permissions(request)
+        response = dict(access="blue" if self.Access.BLUE in access else "red",
```
Comment (ghost) [inline app/gameboard_api.py]: use single quotes

File: static/js/gameboard.js [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    if(redOps===0 && blueOps===0) {
+        stream('GameBoard is only useful if a red or blue operation has been started.')
+    } else {
+        stream('Select red/blue operations to see what the defense detected and responded to.');
+    }
+});
+
+function refresh(){
+    function draw(data){
+        let spacing = '&nbsp;&nbsp;-&nbsp;&nbsp;';
```
Comment (wbooth) [inline static/js/gameboard.js]: if this isn't used we can drop it

File: static/js/gameboard.js [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+function updateExchanges(exchanges, access) {
+    let redPoints = 0;
+    let bluePoints = 0;
+    let pid = 0;
+
+    for(let i = 0; i < exchanges.length; i++){
+        pid = exchanges[i][0]
+        exchange = exchanges[i][1]
+
+        for (let i=0; i<exchange['red'].length; i++){
```
Comment (wbooth) [inline static/js/gameboard.js]: if using I on 193 use a different var here for clarity

File: static/js/gameboard.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    }
+}
+
+function updateExchanges(exchanges, access) {
+    let redPoints = 0;
+    let bluePoints = 0;
+    let pid = 0;
+
+    for(let i = 0; i < exchanges.length; i++){
+        pid = exchanges[i][0]
```
Comment (wbooth) [inline static/js/gameboard.js]: use let to declare the variable here

File: static/js/gameboard.js [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        piece.find('.mid').html(
+            '<span class="gp-pid">'+ pid +'</span>' +
+            '<span class="gp-host">'+ links[0].host +'</span>');
+    }
+}
+
+function updateExchanges(exchanges, access) {
+    let redPoints = 0;
+    let bluePoints = 0;
+    let pid = 0;
```
Comment (wbooth) [inline static/js/gameboard.js]: if pid is not being used here drop let's drop it

File: static/js/gameboard.js [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+    for(let i = 0; i < exchanges.length; i++){
+        pid = exchanges[i][0]
+        exchange = exchanges[i][1]
+
+        for (let i=0; i<exchange['red'].length; i++){
+            redPoints += handOutRedPoints(exchange['red'][i]);
+        }
+        for (let i=0; i<exchange['blue'].length; i++){
```
Comment (wbooth) [inline static/js/gameboard.js]: can we move these to functions since it's duplicated?

**Review by wbooth — CHANGES_REQUESTED** [review]
Code is working and provides a good improvement over the existing code. Can you take a second look to reorganize and clean up code. Looking specifically at duplicated sections (where you are doing the same for red and blue) and breaking up large functions that are doing 2 or 3 different things.

**Review by wbooth — APPROVED** [review]

**Pull request merged**